### DISCUSSION TITLE
ESP-32 Timer test

### DISCUSSION THREAD

**Discussion by ds5apm** [espressif/arduino-esp32#7830] [Q&A]
The first initTimer() runs, and every call to initTimer() from the second onwards displays an error message on the serial monitor. can someone help me?

ESP-32 Timer test
LED_BUILTIN LED ON/OFF:
Serial Monitor 
E (25015) timer_group: timer_isr_callback_add(236): register interrupt service failed
E (28015) timer_group: timer_isr_callback_add(236): register interrupt service failed
E (31015) timer_group: timer_isr_callback_add(236): register interrupt service failed
E (34015) timer_group: timer_isr_callback_add(236): register interrupt service failed
E (37015) timer_group: timer_isr_callback_add(236): register interrupt service failed
.....

////// code /////////////

unsigned long onesec_ms;
hw_timer_t * isrTimer = NULL;
portMUX_TYPE timerMux = portMUX_INITIALIZER_UNLOCKED;

void IRAM_ATTR offTimer(){
  portENTER_CRITICAL_ISR(&timerMux);
  digitalWrite(LED_BUILTIN, !digitalRead(LED_BUILTIN)); 
  portEXIT_CRITICAL_ISR(&timerMux);
};

void initTimer(){
  isrTimer = timerBegin(1, 80, true);
  timerAttachInterrupt(isrTimer, &offTimer, true);
  timerAlarmWrite(isrTimer, 1000000, false); //1sec
  timerAlarmEnable(isrTimer);  
};

void setup() {
  Serial.begin(115200);
  delay(1000);
  pinMode(LED_BUILTIN,OUTPUT);
}

void loop() {
   if ( millis() >= onesec_ms) {
      onesec_ms = millis() + 3000;
      initTimer();      
   }
}

**Comment by mrengineer7777** [comment]
1) You should only begin timer once
2) Why would you attempt to register an interrupt again with timerAttachInterrupt()?

For a simpler Arduino way to do this, use Ticker.  It works great and takes care of the timer setup/teardown.
https://github.com/espressif/arduino-esp32/blob/master/libraries/Ticker/examples/Blinker/Blinker.ino

**Reply by ds5apm** [reply]
The function I want is to start a timer in a specific situation and automatically stop it when the specified time is reached. And, if necessary, initTimer() is called to restart the timer. However, when initTimer() is called again, I did not know the cause of the error, so I found another method.
 
1.timerAlarmWrite(isrTimer, 1000000, true);//reload
2.timerStart(isrTimer); , timerStop(isrTimer); When I use this function, the timer start and stop operation works well.
Thanks for your help.

**Reply by mrengineer7777** [reply]
Glad you found a solution.  Since you are using timerAlarmWrite, watch out for this: https://github.com/espressif/arduino-esp32/issues/7336

**Reply by ds5apm** [reply]
Thanks for introducing me to a good example that I wanted to know about.

**Reply by HomeKidd** [reply]
@ds5apm What was exactly the solution? I'm still facing the same issue.  I need to register an interrupt again with timerAttachInterrupt(), but with different time value. My code is nearly the same as your, first the the ISR task runs perfectly, but after that i have the  `timer_group: timer_isr_callback_add(236): register interrupt service failed` message and the ISR task never runs again

**Comment by murarduino** [comment]
my test code:

TaskHandle_t cpHandle = NULL;
hw_timer_t *timer1 = NULL;

void IRAM_ATTR TimercallBack()
{
 timerStop(timer1);
  Serial.print("Time is UP！");
  Serial.print("  Timer State?: ");
  Serial.println(timerStarted(timer1));
}
void timer_init()
{
  timer1 = timerBegin(0, 40000, true); // Resolution 0.5ms
  timerAttachInterrupt(timer1, TimercallBack, true);
  timerAlarmWrite(timer1, 8 * 2000, true);
  timerAlarmEnable(timer1);

  // timerAlarmEnable(timer1);
}



void cp_task(void *ptr)
{
  timer_init();
  timerStop(timer1);

  while (1)
  {
    if ( !timerStarted(timer1))
    {
    //  timer_init();
    timerStart(timer1);
      Serial.println("Time Start/Restart!");
    }


    Serial.print(xTaskGetTickCount());
    Serial.print("  Timer State?: ");
    Serial.println(timerStarted(timer1));
    vTaskDelay(1000);

  }
}

void setup() {
  // put your setup code here, to run once:
  Serial.begin(115200);
  xTaskCreate(cp_task, "CP_task", 1024 * 3, NULL, 1, &cpHandle);
  // Serial.println("Hello, ESP32!");
  vTaskDelete(NULL);
}

void loop() {

}
